### PR TITLE
coprocessor: move ExecSummaryCollector out of Executors into a wrapper struct

### PR DESCRIPTION
**Pull request by kennytm**
Signed-off-by: kennytm <kennytm@gmail.com>

## What have you changed? (mandatory)

Removed the `ExecSummaryCollector` field from *all* Executors and BatchExecutors. Instead, introduced a `WithSummaryCollector<C, T>` struct as a wrapper of the concrete executors, and implement the traits on it.

The main purpose is to factor out the repetition of the pattern `on_start_iterate() ... on_finish_iterate()`.

Note: Currently `WithSummaryCollector` will _always_ execute `summary_collector.collect_into()` inside `collect_statistics()`, but it was not always the case in the original implementation. I'm not sure if this is dedicated or just an oversight. We could add a flag to control this behavior if the omission is indeed dedicated.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

```
cargo test -- coprocessor
```

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

**Comment by kennytm** [comment]
PTAL @breeswish @lonng

**Comment by breeswish** [comment]
Thanks a lot! I will merge this PR after hash agg is being merged to avoid causing too much conflicts with https://github.com/tikv/tikv/pull/3898

**Comment by breeswish** [comment]
Hold for https://github.com/tikv/tikv/pull/4752

File: src/coprocessor/dag/executor/aggregation.rs [hunk 1]
```diff
@@ -241,16 +237,9 @@ impl<C: ExecSummaryCollector> HashAggExecutor<C> {
...
     fn next(&mut self) -> Result<Option<Row>> {
-        let timer = self.inner.summary_collector.on_start_iterate();
-        let ret = self.next_impl();
-        if let Ok(Some(_)) = ret {
-            self.inner.summary_collector.on_finish_iterate(timer, 1)
-        } else {
-            self.inner.summary_collector.on_finish_iterate(timer, 0)
-        }
-        ret
+        self.next_impl()
```
Comment (breeswish) [inline src/coprocessor/dag/executor/aggregation.rs]: We can remove next_impl now.

File: src/coprocessor/dag/executor/aggregation.rs [hunk 2]
```diff
@@ -274,16 +263,9 @@ impl<C: ExecSummaryCollector> Executor for HashAggExecutor<C> {
...
```
Comment (breeswish) [inline src/coprocessor/dag/executor/aggregation.rs]: We can remove next_impl now.

File: src/coprocessor/dag/executor/scan.rs [hunk 3]
```diff
@@ -152,16 +149,9 @@ impl<C: ExecSummaryCollector, S: Store, T: InnerExecutor> ScanExecutor<C, S, T>
...
     fn next(&mut self) -> Result<Option<Row>> {
-        let timer = self.summary_collector.on_start_iterate();
-        let ret = self.next_impl();
-        if let Ok(Some(_)) = ret {
-            self.summary_collector.on_finish_iterate(timer, 1);
-        } else {
-            self.summary_collector.on_finish_iterate(timer, 0);
-        }
-        ret
+        self.next_impl()
```
Comment (breeswish) [inline src/coprocessor/dag/executor/scan.rs]: We can remove next_impl now.

File: src/coprocessor/dag/executor/scan.rs [hunk 4]
```diff
@@ -186,9 +176,7 @@ impl<C: ExecSummaryCollector, S: Store, T: InnerExecutor> Executor for ScanExecu
         }
     }
 
-    fn collect_execution_summaries(&mut self, target: &mut [ExecSummary]) {
-        self.summary_collector.collect_into(target);
-    }
+    fn collect_execution_summaries(&mut self, _: &mut [ExecSummary]) {}
```
Comment (breeswish) [inline src/coprocessor/dag/executor/scan.rs]: maybe we can provide a default impl for this function (for normal executor).

File: src/coprocessor/dag/executor/selection.rs [hunk 5]
```diff
@@ -57,16 +54,9 @@ impl<C: ExecSummaryCollector> SelectionExecutor<C> {
...
     fn next(&mut self) -> Result<Option<Row>> {
-        let timer = self.summary_collector.on_start_iterate();
-        let ret = self.next_impl();
-        if let Ok(Some(_)) = ret {
-            self.summary_collector.on_finish_iterate(timer, 1);
-        } else {
-            self.summary_collector.on_finish_iterate(timer, 0);
-        }
-        ret
+        self.next_impl()
```
Comment (breeswish) [inline src/coprocessor/dag/executor/selection.rs]: We can remove next_impl now.

File: src/coprocessor/dag/executor/topn.rs [hunk 6]
```diff
@@ -120,16 +117,9 @@ impl<C: ExecSummaryCollector> TopNExecutor<C> {
...
     fn next(&mut self) -> Result<Option<Row>> {
-        let timer = self.summary_collector.on_start_iterate();
-        let ret = self.next_impl();
-        if let Ok(Some(_)) = ret {
-            self.summary_collector.on_finish_iterate(timer, 1);
-        } else {
-            self.summary_collector.on_finish_iterate(timer, 0);
-        }
-        ret
+        self.next_impl()
```
Comment (breeswish) [inline src/coprocessor/dag/executor/topn.rs]: We can remove next_impl now.

File: src/coprocessor/dag/batch/executors/selection_executor.rs [hunk 7]
```diff
@@ -124,18 +115,12 @@ impl<C: ExecSummaryCollector, Src: BatchExecutor> BatchExecutor for BatchSelecti
 
     #[inline]
     fn next_batch(&mut self, scan_rows: usize) -> BatchExecuteResult {
-        let timer = self.summary_collector.on_start_iterate();
-        let result = self.handle_next_batch(scan_rows);
-        self.summary_collector
-            .on_finish_iterate(timer, result.data.rows_len());
-        result
+        self.handle_next_batch(scan_rows)
```
Comment (breeswish) [inline src/coprocessor/dag/batch/executors/selection_executor.rs]: maybe we can remove handle_next_batch

File: src/coprocessor/dag/executor/scan.rs (hunk 4 above)
Comment (kennytm) [reply src/coprocessor/dag/executor/scan.rs]: All other impls forward to call `self.inner.collect_execution_summaries(target)`, but for this to be defaultable the Executor trait would need to know there may be an `inner` 🤔

**Comment by kennytm** [comment]
/test

Wow two simultaneous spurious failures.

1. [`raftstore::test_lease_read::test_read_index_when_transfer_leader`](https://internal.pingcap.net/idc-jenkins/blue/rest/organizations/jenkins/pipelines/tikv_ghpr_test/runs/4847/nodes/130/log/?start=0)
2. [`pd::test_rpc_client::test_restart_leader_insecure`](https://internal.pingcap.net/idc-jenkins/blue/rest/organizations/jenkins/pipelines/tikv_ghpr_test/runs/4847/nodes/133/log/?start=0)

**Comment by kennytm** [comment]
PTAL @AndreMouche

**Comment by breeswish** [comment]
@lonng Yes. I think we can later cheery pick multiples altogether, not necessary to cherry pick one by one.

**Pull request merged**